onyo/lib/tests/test_git.py

```diff
@@ -67,283 +87,245 @@ def test_GitRepo_find_root(tmp_path: Path) -> None:
-
-    # test that GitRepo.commit() raises a ValueError if no commit message is
-    # provided
-    with pytest.raises(ValueError):
-        new_git.commit()
+    assert test_file not in gitrepo.files
+    # fresh repo w/ no commit
+    assert gitrepo.get_hexsha() is None
+    gitrepo.commit(test_file, message="Create file for test")
+    hexsha = gitrepo.get_hexsha()
+    # created a commit:
+    assert hexsha is not None
+    # only one commit (HEAD~1 does not exist):
+    pytest.raises(ValueError, gitrepo.get_hexsha, 'HEAD~1')
+    assert test_file in gitrepo.files
 
     # modify an existing file, and add it
     test_file.open('w').write('Test: content')
-    new_git.add(test_file)
-    assert test_file in new_git.files_staged
-
-    # commit staged changes
-    new_git.commit("Test commit message")
-    assert test_file in new_git.files
-
-    # after files are `GitRepo.commit()`ed they should not be cached in the
-    # properties GitRepo.files_changed or GitRepo.files_staged anymore
-    assert test_file not in new_git.files_changed
-    assert test_file not in new_git.files_staged
-
-
-def test_GitRepo_stage_and_commit(tmp_path: Path) -> None:
-    """
-    `GitRepo.stage_and_commit()` must allow to add+commit changed files.
-
-    This test follows the scheme of `test_GitRepo_add()`.
-    """
-    subprocess.run(['git', 'init', tmp_path])
-    new_git = GitRepo(tmp_path)
-    test_file = new_git.root / 'test_file.txt'
-
-    # add a file
-    test_file.open('w').write('Test: content')
-    assert test_file in new_git.files_untracked
-
-    # add+commit a changed file
-    new_git.stage_and_commit(test_file, "Test commit message")
-    assert test_file in new_git.files
-
-    # after files are `GitRepo.stage_and_commit()`ed they should not be cached
-    # in the properties GitRepo.files_changed or GitRepo.files_staged
-    assert test_file not in new_git.files_untracked
-    assert test_file not in new_git.files_changed
-    assert test_file not in new_git.files_staged
+    gitrepo.commit(test_file, "Test commit message")
+    assert hexsha == gitrepo.get_hexsha('HEAD~1')
+    assert test_file in gitrepo.files
+
+
+@pytest.mark.gitrepo_contents((Path('some.file'),
+                               "some content"),
+                              (Path('top') / 'mid' / "another.txt",
+                               "")
+                              )
+def test_GitRepo_get_subtrees(gitrepo) -> None:
+    # add an untracked files:
+    for d in gitrepo.test_annotation['directories']:
+        untracked = d / "untracked"
+        untracked.touch()
+
+    # only returns tracked files underneath the given directory:
+    for d in gitrepo.test_annotation['directories']:
+        tree = gitrepo.get_subtrees([d])
+        assert [p for p in gitrepo.test_annotation['files'] if d in p.parents] == tree
+
+    # defaults to the entire worktree:
+    assert [p for p in gitrepo.test_annotation['files']] == gitrepo.get_subtrees()
+
+    # several dirs:
```

Comment:
This block tests how `get_subtrees` behave when several dirs are passed to it.
To do that, it makes sure that the `gitrepo` actually has several dirs to look into. Currently there's only one particular repo this test function is running on - as defined by the `gitrepo_content` marker. However, we could parameterize that and throw a whole bunch of different repos at that test function. And then it would run that part of the test only the repos where it actually makes sense. If the repo only has one dir, there's no point it trying to test this behavior. Hence, `len(gitrepo.test_annotation['directories']) > 1` and nothing to test for `len(...) == 1`. If there's only one dir in that repo, we are done testing. Lines 222-224 deal with passing a single dir already.